Instantiating struct with global on heap causes c error: expected struct or union but not 'struct main__GameObject *'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 5c65e58.9300982
OS: windows, Microsoft Windows 11 Pro v22631 64-bit
Processor: 24 cpus, 64bit, little endian, 

get_current_dir: D:\vbug
vexe: C:\v\v.exe
vexe mtime: 2024-11-25 10:39:04

vroot: OK, value: C:\v
VMODULES: OK, value: C:\Users\info\.vmodules
VTMP: OK, value: C:\Users\info\AppData\Local\Temp\v_0

Git version: git version 2.42.0.windows.2
Git vroot status: weekly.2024.47-35-g93009823-dirty
.git/config present: true

CC version: Error: 'cc' is not recognized as an internal or external command,

operable program or batch file.


emcc version: Error: 'emcc' is not recognized as an internal or external command,

operable program or batch file.


thirdparty/tcc status: thirdparty-windows-amd64 b425ac82
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg src/main.v`

```v
module main

@[heap]
struct Game {
mut:
	object_id u32
}

__global (
	g Game
)

@[heap]
struct GameObject  {
mut:
	id        int = g.object_id++
}

fn main() {
	mut g := GameObject{}
	eprintln("G=${g}")
}
```

**What did you expect to see?**

no c error

**What did you see instead?**

```
$ /c/v/v -enable-globals run .
================== C compilation error (from tcc): ==============
cc: C:/Users/info/AppData/Local/Temp/v_0/vbug.01JDHEFNZ9FK39SECZ12MT17SW.tmp.c:7194: warning: implicit declaration of function 'tcc_backtrace'
cc: C:/Users/info/AppData/Local/Temp/v_0/vbug.01JDHEFNZ9FK39SECZ12MT17SW.tmp.c:13895: error: expected struct or union but not 'struct main__GameObject *'
... (the original output was 3 lines long, and was truncated to 2 lines)
=================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error:
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzQ0NTY0MmQyZWY0Y2JjYzQ2OGQxNjkiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.lfs0QprbH0ZuaMc7_FjE9NbXdWQyU-Y3oxVqPrv13x0">Huly&reg;: <b>V_0.6-21411</b></a></sub>

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -141,7 +141,7 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 			}
 		} else if mut left is ast.Ident && node.op == .decl_assign {
 			if left.name in c.global_names {
-				c.error('already exists a global variable named `${left.name}`', left.pos)
+				c.error('the global variable named `${left.name}` already exists', left.pos)
 			}
 		}
 		is_blank_ident := left.is_blank_ident()

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -141,7 +141,7 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 			}
 		} else if mut left is ast.Ident && node.op == .decl_assign {
 			if left.name in c.global_names {
-				c.error('the global variable named `${left.name}` already exists', left.pos)
+				c.note('the global variable named `${left.name}` already exists', left.pos)
 			}
 		}
 		is_blank_ident := left.is_blank_ident()

--- a/vlib/v/checker/assign.v
+++ b/vlib/v/checker/assign.v
@@ -139,6 +139,10 @@ fn (mut c Checker) assign_stmt(mut node ast.AssignStmt) {
 				c.error('cannot reassign using range expression on the left side of an assignment',
 					left.pos)
 			}
+		} else if mut left is ast.Ident && node.op == .decl_assign {
+			if left.name in c.global_names {
+				c.error('already exists a global variable named `${left.name}`', left.pos)
+			}
 		}
 		is_blank_ident := left.is_blank_ident()
 		mut left_type := ast.void_type

--- a/vlib/v/checker/tests/globals/global_var_redeclare.out
+++ b/vlib/v/checker/tests/globals/global_var_redeclare.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/globals/global_var_redeclare.vv:4:2: notice: the global variable named `foobar` already exists
+    2 | 
+    3 | fn main() {
+    4 |     foobar := 2
+      |     ~~~~~~
+    5 |     println(foobar)
+    6 | }

--- a/vlib/v/checker/tests/globals/global_var_redeclare.vv
+++ b/vlib/v/checker/tests/globals/global_var_redeclare.vv
@@ -1,0 +1,6 @@
+__global foobar = 1
+
+fn main() {
+	foobar := 2
+	println(foobar)
+}

